src/pepr/operator/controllers/keycloak/authservice/authservice.ts

```diff
@@ -79,45 +92,97 @@ export async function authservice(pkg: UDSPackage, clients: Map<string, Client>)
+
+  // Return the new status objects for status update
+  return newAuthserviceClients;
 }
 
 export async function purgeAuthserviceClients(
   pkg: UDSPackage,
-  newAuthserviceClients: string[] = [],
-  isAmbient = false,
-) {
-  R.difference(pkg.status?.authserviceClients || [], newAuthserviceClients).forEach(
-    async clientId => {
-      log.info(`Removing stale authservice chain for client ${clientId}`);
-      // Use client ID directly as the waypoint ID
-      const waypointId = clientId;
-      // Get the full waypoint name for authorization policies
-      const fullWaypointName = getWaypointName(waypointId);
+  newAuthserviceClients: AuthserviceClient[] = [],
+  previousMeshMode: string,
+  currentMeshMode: string,
+): Promise<void> {
+  const isAmbient = shouldUseAmbientWaypoint(pkg);
+  const prevClients = pkg.status?.authserviceClients || [];
+
+  // Check if mesh mode changed
+  const meshModeChanged = previousMeshMode !== currentMeshMode;
+
+  // Find clients to clean up:
+  // 1. Removed clients
+  // 2. Clients with changed selectors
+  // 3. All clients if mesh mode changed
+  const clientsToCleanup = meshModeChanged
+    ? prevClients // Clean up all previous clients if mesh mode changed
+    : prevClients.filter(oldClient => {
+        // Find matching client by ID
+        const match = newAuthserviceClients.find(
+          newClient => newClient.clientId === oldClient.clientId,
+        );
+
+        // If no match found, this client was removed
+        if (!match) {
+          return true;
+        }
+
+        // Check if either selector is empty
+        const oldSelectorEmpty =
+          !oldClient.selector || Object.keys(oldClient.selector).length === 0;
+        const newSelectorEmpty = !match.selector || Object.keys(match.selector).length === 0;
+
+        // Both empty, no need for cleanup
+        if (oldSelectorEmpty && newSelectorEmpty) {
+          return false;
+        } else if (!oldSelectorEmpty && !newSelectorEmpty) {
+          // Both non-empty, check if they're different
+          return !R.equals(match.selector, oldClient.selector);
+        } else if (oldSelectorEmpty && !newSelectorEmpty) {
+          // Changed from empty to non-empty
+          return true;
+        } else {
+          // Changed from non-empty to empty
+          return false;
+        }
+      });
+
+  if (meshModeChanged) {
+    log.info(
+      `Mesh mode changed from ${previousMeshMode} to ${currentMeshMode}, cleaning up waypoint labels`,
+    );
+  }
+
+  await Promise.all(
+    clientsToCleanup.map(async client => {
+      const fullWaypointName = getWaypointName(client.clientId);
+      log.info(`Cleaning up authservice client ${client.clientId}`, {
+        reason: meshModeChanged ? "mesh_mode_change" : "client_removed_or_modified",
+      });
+
       await reconcileAuthservice(
-        { name: clientId, action: Action.RemoveClient },
+        { name: client.clientId, action: Action.RemoveClient },
```

Comment:
This results in the authservice chain getting removed if the selector changed. I think we need to make sure that `reconcileAuthservice` is only called when the client no longer exists, not when a selector/mesh mode changes.

Reply:
Looking at the refactor - the structure and logic makes sense to me. I'm a bit confused looking at the flow of when `purgeAuthserviceClients` is called though:
1. Reconciler calls `authservice`.
2. That calls `reconcileAuthservice` for each client in the Package CR (new or updated).
3. Then we call `purgeAuthserviceClients`.
4. That calls `reconcileAuthservice` w/ remove action for each client that has been removed.
5. Also calls `reconcileAuthservice` w/ add action for each client that has been updated.

2 and 5 seem redundant here, and I think we might be missing some logic to cleanup labels for 5? Going to deploy and validate how this is working though.